engine/verification/finder/engine.go

```diff
@@ -326,52 +355,202 @@ func (e *Engine) checkCachedReceipts() {
 				return
 			}
 
-			// checks if the result has already been handled
+			// checks if the result has already been processed or discarded
 			if e.processedResultIDs.Has(resultID) {
 				log.Debug().Msg("drops handling already processed result")
 				return
 			}
+			if e.discardedResultIDs.Has(resultID) {
+				log.Debug().Msg("drops handling already discarded result")
+				return
+			}
 
-			// adds receipt to pending or ready mempools depending on its processable status
 			ready := e.isProcessable(&rdp.Receipt.ExecutionResult)
-			if ready {
-				// block for the receipt is available,
-				// receipt is ready for process
-				ok := e.readyReceipts.Add(rdp)
-				if !ok {
-					log.Debug().Msg("drops adding duplicate receipt to ready mempool")
-					return
-				}
-			} else {
-				// block for the receipt is not available
-				// receipt moves to pending mempool
-				ok := e.pendingReceipts.Add(rdp)
-				if !ok {
-					log.Debug().Msg("drops adding duplicate receipt to pending mempool")
-					return
-				}
-
-				// marks receipt pending for its block ID
-				err := e.pendingReceiptIDsByBlock.Append(rdp.Receipt.ExecutionResult.BlockID, receiptID)
-				if err != nil {
-					e.log.Error().
-						Err(err).
-						Hex("block_id", logging.ID(rdp.Receipt.ExecutionResult.BlockID)).
-						Hex("receipt_id", logging.ID(receiptID)).
-						Msg("could not append receipt to receipt-ids-by-block mempool")
-				}
+			if !ready {
+				// adds receipt to pending mempool
+				e.addToPending(rdp)
+				return
 			}
 
-			// records the execution receipt id based on its result id
-			err := e.receiptIDsByResult.Append(resultID, receiptID)
-			if err != nil {
-				log.Debug().Err(err).Msg("could not add receipt id to receipt-ids-by-result mempool")
+			// adds receipt to ready mempool
+			e.addToReady(rdp)
+		}()
+	}
+}
+
+// addToReady encapsulates the logic around adding a ReceiptDataPack to ready receipts mempool.
+// The ReceiptDataPack is however discarded it if finder engine is not staked at its block id.
+func (e *Engine) addToReady(receiptDataPack *verification.ReceiptDataPack) {
+	receiptID := receiptDataPack.Receipt.ID()
+	resultID := receiptDataPack.Receipt.ExecutionResult.ID()
+	blockID := receiptDataPack.Receipt.ExecutionResult.BlockID
+	log := e.log.With().
+		Hex("origin_id", logging.ID(receiptDataPack.OriginID)).
+		Hex("block_id", logging.ID(blockID)).
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).Logger()
+
+	// checks whether verification node is staked at snapshot of this result's block.
+	ok, err := e.stakedAtBlockID(blockID)
+	if err != nil {
+		log.Debug().Err(err).Msg("could verify stake of verification node for result")
```

Comment:
the above code snippet is part of `addToReady` method on the `finder` engine. This method is get called once `finder` realizes that the block for the [receipt is available in the `blocks` storage](https://github.com/onflow/flow-go/blob/da958f7bc9860f43195fbb9136d221cc5c69629f/engine/verification/finder/engine.go#L368-L376). Since `storage` and  `state` have the same backend database, I think it can never be the case that the block is not found at this stage, since its existence has already been checked before invoking `addToReady`.